### DISCUSSION TITLE
TS SDK pain points

### DISCUSSION THREAD

**Discussion by matt-user** [FuelLabs/fuels-ts#1881] [Ideas]
Hey all, I finished working on a frontend using the ts-sdk and thought I would share some pain points.

 1. [ ] #1882
 1. Nit: fuels dev docs should mention build, type gen, and deploy (I think it only mentions build and deploy)
 1. when I used fuels init without create fuels I had difficulty using fuels dev "default account cannot deploy contracts bc of insufficient funds"
 1. I didn't use `create fuels` bc I didn't see it in the cli portion of the docs
 1. I also had issues with the private key field not working on the fuels config.
 1. Typegen does not automatically use configured inputs and outputs specified in the config
 1. Difficulty importing enum from contract Abi so I had to remember what the underlying data of the enum
 1. Its hard to distinguish log events from each other.  I have to look at the sway code.
 https://github.com/FuelLabs/fuels-ts/issues/2330
 1. Add some documentation about the .fuels folder generated (just a note that it gets created from fuels is probably sufficient)
 1. [ ] #1886 
![Screenshot 2024-03-14 at 9 52 46 AM](https://github.com/FuelLabs/fuels-ts/assets/54373384/488a044f-29c8-4f22-9afe-f79696c484f2)
 1. It would also be nice to just have to list one connector and not all of them

**Comment by arboleya** [comment]
@matt-user Thanks for putting this feedback together.

I did some formatting and started addressing the first point.

**Comment by arboleya** [comment]
@matt-user How often the 10th item happens?

**Reply by matt-user** [reply]
It happened every time for me.  I was not able to get it to work.

**Comment by danielbate** [comment]
> Difficulty importing enum from contract Abi so I had to remember what the underlying data of the enum

Do you have an example for this? We can definitely improve enums, #1844 suggests something similar so may be able to sync up the solution.

**Reply by matt-user** [reply]
Looking back my issue was actually importing the enums not constructing them.  I'd like to do something like this `gameState === StateOutput.Ended`, but I can't import `StateOutput` so instead I have to do `gameState === "Ended"`. It could be an issue in how I build my app I guess
<img width="1023" alt="Screenshot 2024-03-27 at 1 54 41 PM" src="https://github.com/FuelLabs/fuels-ts/assets/54373384/3e909a52-917d-4479-9a33-ba4519b530ca">

**Comment by petertonysmith94** [comment]
### Investigate further

> 3. when I used fuels init without create fuels I had difficulty using fuels dev "default account cannot deploy contracts bc of insufficient funds"

> 5.  I also had issues with the private key field not working on the fuels config.

Potentially related to:

- https://github.com/FuelLabs/fuels-ts/issues/2783

---

### Raise issues

> 2. Nit: fuels dev docs should mention build, type gen, and deploy (I think it only mentions build and deploy)

Missing links on the [homepage](https://docs.fuel.network/docs/fuels-ts/fuels-cli/) to:
  - `fuels node`
  - `fuels typegen`
  - `fuels versions`

> 9. Add some documentation about the .fuels folder generated (just a note that it gets created from fuels is probably sufficient)

Add issue around documenting the `.fuels` folder

---

### Further discussion required

> 6. Typegen does not automatically use configured inputs and outputs specified in the config

I can see that we currently don't pass down the config to the typegen command:

https://github.com/FuelLabs/fuels-ts/blob/master/packages/fuels/src/cli.ts#L94-L96

However, potentially the command you were looking to use was the `fuels build` command which will build the Sway code and typegen. Maybe this is a documentation issue? 

> 7. Difficulty importing enum from contract Abi so I had to remember what the underlying data of the enum

We do export types for the enums in the contract (example below).

```sway
pub enum Difficulty {
    Easy: bool,
    Medium: bool,
    Hard: bool,
}
```

Generates:

```ts
export type DifficultyInput = Enum<{
  Easy: boolean;
  Medium: boolean;
  Hard: boolean;
}>;
```

Maybe we also want to build out the TypeScript base enums as well?

```ts
export enum Difficulty {
  Easy = "Easy",
  Medium = "Medium",
  Hard = "Hard",
}
```

What do you think @nedsalk?

> 8. Its hard to distinguish log events from each other. I have to look at the sway code.

:orange: There is this [issue](https://github.com/FuelLabs/fuels-ts/issues/2330) open for supporting logs in Typegen. I know currently all our logs are typed as `any` which is not ideal.

Do you think we can alleviate this with the above? cc. @nedsalk

---

### Maybe resolved

> 4. I didn't use `create fuels` bc I didn't see it in the cli portion of the docs

I believe this has been resolved by moving [`Creating a Fuel dApp`](https://docs.fuel.network/docs/fuels-ts/creating-a-fuel-dapp/) after `Getting Started`.

What do you think @matt-user?

> 11. It would also be nice to just have to list one connector and not all of them

:green: I assume this has been resolved by the `defaultConnectors` exports from `@fuels/connectors` - would you agree @matt-user?

```ts
import { Fuel } from "fuels";
import { defaultConnectors } from "@fuels/connectors";

const fuel = new Fuel({
  connectors: defaultConnectors({ devMode: true }),
});
```

**Reply by nedsalk** [reply]
> Maybe we also want to build out the TypeScript base enums as well?
> ```ts
>export enum Difficulty {
>  Easy = "Easy",
>  Medium = "Medium",
>  Hard = "Hard",
>}```

We can do that when mapping native sway enums (that only have `()` types), but generally enums behave the same as structs so we can't do much about that and we have to map them as types and not as typescript `enum`s.

>Its hard to distinguish log events from each other. I have to look at the sway code.
> 🍊 There is this https://github.com/FuelLabs/fuels-ts/issues/2330 open for supporting logs in Typegen. I know currently all our logs are typed as any which is not ideal.
Do you think we can alleviate this with the above?

Yes, that issue should fix it.
